### ISSUE TITLE
Base class conversion method is repeated in child class definition

### ISSUE THREAD

**Issue by dogancan** [google/clif#19]
After yesterday's update, rudimentary wrappers I had previously written for the C++ iostream and stringstream types stopped working. For some reason CLIF now generates the basic_stream conversion methods twice.

I was able to replicate the situation in this simplified setup.

```c++
#ifndef TEST_H_
#define TEST_H_

template <class T>
class base2 {
public:
  bool good() { return true; }
};

template <class T>
class derived1 : virtual public base2<T> {};

template <class T>
class derived2 : virtual public base2<T> {};

template <class T>
class derived3 : public derived1<T>, public derived2<T> {};

typedef derived3<char> derived3_char;

#endif
```
CLIF wrapper:
```python
from "test.h":
  class derived3_char:
    def good(self) -> bool
```
The definition for as_base2_char is repeated in the CLIF output (.cc).
```c++
...

// Implicit cast this as ::derived1<char>*
static PyObject* as_derived1_char(PyObject* self) {
  ::derived1<char>* p = ::clif::python::Get(reinterpret_cast<wrapper*>(self)->cpp)
;
  if (p == nullptr) return nullptr;
  return PyCapsule_New(p, C("::derived1<char>"), nullptr);
}

// Implicit cast this as ::derived2<char>*
static PyObject* as_derived2_char(PyObject* self) {
  ::derived2<char>* p = ::clif::python::Get(reinterpret_cast<wrapper*>(self)->cpp);
  if (p == nullptr) return nullptr;
  return PyCapsule_New(p, C("::derived2<char>"), nullptr);
}

// Implicit cast this as ::base2<char>*
static PyObject* as_base2_char(PyObject* self) {
  ::base2<char>* p = ::clif::python::Get(reinterpret_cast<wrapper*>(self)->cpp);
  if (p == nullptr) return nullptr;
  return PyCapsule_New(p, C("::base2<char>"), nullptr);
}

// Implicit cast this as ::base2<char>*
static PyObject* as_base2_char(PyObject* self) {
  ::base2<char>* p = ::clif::python::Get(reinterpret_cast<wrapper*>(self)->cpp);
  if (p == nullptr) return nullptr;
  return PyCapsule_New(p, C("::base2<char>"), nullptr);
}

...
```
~It seems this issue arises only in the case of virtual inheritance.~

**Comment by mrovner** [comment]
To solve a diamond inheritance problem one should use [virtual inheritance](https://www.cprogramming.com/tutorial/virtual_inheritance.html)

**Comment by dogancan** [comment]
I looked at this again and it seems virtual inheritance is not related to this issue. 

Extra definition is produced only in the case of class template instantiation. Without the template parameters, CLIF produces a single conversion method for the root class.

**Comment by mrovner** [comment]
Thank you for the report. I checked in a workaround and we'll investigate further.

**Comment by dogancan** [comment]
Thanks.

**Comment by zhongming1993** [comment]
Fixed and will be released in (coming) version 0.3.1.

**Comment by rwgk** [comment]
This repo was unmaintained from December 2017 to September 2019.                
It was recently updated, see https://github.com/google/clif/discussions/36#discussioncomment-85120 for some background.
                                                                                
If this issue is still relevant please repopen it or ask questions at https://github.com/google/clif/discussions/.
